File: web/restore_rules_spanish.mjs
/**
 * restore_rules_spanish.mjs
 * Regenera en español las 130 normas editoriales eliminadas.
 * Usa Gemini para generar nombre+descripción en español a partir del nombre inglés original.
 *
 * Uso: node restore_rules_spanish.mjs [--dry-run]
 */

import { initializeApp, cert } from "firebase-admin/app";
import { getFirestore, FieldValue } from "firebase-admin/firestore";
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";
import { GoogleGenAI } from "@google/genai";

const __dirname = dirname(fileURLToPath(import.meta.url));
const DRY_RUN = process.argv.includes("--dry-run");

// ── Firebase Admin ──────────────────────────────────────────
const saPath = join(__dirname, "../ai-worker/service-account.json");
const serviceAccount = JSON.parse(readFileSync(saPath, "utf8"));
initializeApp({ credential: cert(serviceAccount) });
const db = getFirestore();

// ── Gemini ───────────────────────────────────────────────────
// Load API key from .env.local
let GEMINI_API_KEY = "";
try {
  const envContent = readFileSync(join(__dirname, ".env.local"), "utf8");
  const match = envContent.match(/GEMINI_API_KEY\s*=\s*(.+)/);
  if (match) GEMINI_API_KEY = match[1].trim();
} catch { /* optional */ }

// Try process env as fallback
if (!GEMINI_API_KEY) GEMINI_API_KEY = process.env.GEMINI_API_KEY ?? "";

if (!GEMINI_API_KEY) {
  console.error("❌ GEMINI_API_KEY no encontrada en .env.local ni en el entorno.");
  process.exit(1);
}

const ai = new GoogleGenAI({ apiKey: GEMINI_API_KEY });

// ── Org to restore into ──────────────────────────────────────
const ORG_ID = "G8rTqFMuH1xFACIsjEsV";
const ORG_NAME = "Biblioteca Homo Legens";

// ── Deleted rules (from dry-run output) ────────────────────
// These were the ACTIVE rules in English that got deleted.
// Each entry: { englishName, source, category? }
const DELETED_RULES = [
  { englishName: "Italics: Foreign words and Latin phrases", source: "Nosotros", category: "typography" },
  { englishName: "Copyright Page Inclusion", source: "Nosotros", category: "format" },
  { englishName: "Quotation Marks: Primary", source: "Nosotros", category: "typography" },
  { englishName: "Thousand Separator for Years", source: "Nosotros", category: "style" },
  { englishName: "Consistent Numbering Style", source: "Nosotros", category: "style" },
  { englishName: "Italics for Foreign Words and Titles", source: "Conservatismo", category: "typography" },
  { englishName: "Capitalization of Religious Terms", source: "Conservatismo", category: "grammar" },
  { englishName: "Font Style", source: "Conservatismo", category: "format" },
  { englishName: "Spelling Correction: 'umanidad'", source: "Conservatismo", category: "grammar" },
  { englishName: "Chapter Headings Alignment", source: "Entre los impostores", category: "format" },
  { englishName: "Chapter Title Formatting", source: "Entre los impostores", category: "format" },
  { englishName: "Consistent Spacing", source: "Entre los impostores", category: "format" },
  { englishName: "Verb Choice for 'marcha'", source: "Conservatismo", category: "style" },
  { englishName: "Chapter Title Formatting", source: "Nosotros", category: "format" },
  { englishName: "Final Page Inclusion", source: "Nosotros", category: "format" },
  { englishName: "Em Dashes for Parenthetical Phrases", source: "Conservatismo", category: "typography" },
  { englishName: "Block Quotes for Extended Passages", source: "Entre los impostores", category: "format" },
  { englishName: "Punctuation: Guillemets for quotes", source: "Nosotros", category: "typography" },
  { englishName: "Spelling Correction: 'aqullos'", source: "Conservatismo", category: "grammar" },
  { englishName: "Punctuation: Em dashes", source: "Conservatismo", category: "typography" },
  { englishName: "Consistency in Naming Conventions", source: "Entre los impostores", category: "style" },
  { englishName: "Em-Dashes for Parentheticals", source: "Entre los impostores", category: "typography" },
  { englishName: "Noun Choice for 'cole'", source: "Conservatismo", category: "style" },
  { englishName: "Colophon Formatting", source: "Entre los impostores", category: "format" },
  { englishName: "Spelling Correction: 'istoria'", source: "Conservatismo", category: "grammar" },
  { englishName: "Sentence Structure and Flow", source: "Entre los impostores", category: "style" },
  { englishName: "English Double Quotes for Secondary Quotes", source: "Nosotros", category: "typography" },
  { englishName: "Consistency in Acronyms/Abbreviations", source: "Conservatismo", category: "style" },
  { englishName: "Capitalization: 'Estado'", source: "Conservatismo", category: "grammar" },
  { englishName: "Font Style: Body Text", source: "Entre los impostores", category: "format" },
  { englishName: "Footnote Style", source: "Entre los impostores", category: "format" },
  { englipshName: "Ellipsis Spacing", source: "Entre los impostores", category: "typography" },
  { englishName: "Spelling Correction: 'rogresismo'", source: "Conservatismo", category: "grammar" },
  { englishName: "Typo Correction: Missing 'd' or 'g'", source: "Conservatismo", category: "grammar" },
  { englishName: "Imprint Page Details", source: "Nosotros", category: "format" },
  { englishName: "Table of Contents Inclusion", source: "Nosotros", category: "format" },
  { englishName: "Comma Usage for Clarity", source: "Entre los impostores", category: "grammar" },
  { englishName: "Use of Spanish Guillemets for Quotes", source: "Conservatismo", category: "typography" },
  { englishName: "English Single Quotes for Tertiary Quotes", source: "Nosotros", category: "typography" },
  { englishName: "Logo Placement", source: "Entre los impostores", category: "format" },
  { englishName: "Page Numbering", source: "Entre los impostores", category: "format" },
  { englishName: "Comma Usage for Parenthetical Phrases", source: "Conservatismo", category: "grammar" },
  { englishName: "Footnote Numbering", source: "Entre los impostores", category: "format" },
  { englishName: "Punctuation: Semicolons", source: "Nosotros", category: "typography" },
  { englishName: "Punctuation: Periods", source: "Entre los impostores", category: "typography" },
  { englishName: "Legal Disclaimer", source: "Nosotros", category: "format" },
  { englishName: "Quotation Marks: Nested/Specific Terms", source: "Conservatismo", category: "typography" },
  { englishName: "Dialogue Punctuation", source: "Entre los impostores", category: "typography" },
  { englishName: "Dialogue Punctuation", source: "Nosotros", category: "typography" },
  { englishName: "Em Dash Usage in Dialogue", source: "Conservatismo", category: "typography" },
  { englishName: "Content Expansion/Clarification", source: "Entre los impostores", category: "style" },
  { englishName: "Quotation Marks Usage", source: "Conservatismo", category: "typography" },
  { englishName: "Paragraph Indentation", source: "Entre los impostores", category: "format" },
  { englishName: "Pronoun Agreement", source: "Conservatismo", category: "grammar" },
  { englishName: "Punctuation: Ellipses", source: "Nosotros", category: "typography" },
  { englishName: "Typo Correction: 'onservatismo'", source: "Conservatismo", category: "grammar" },
  { englishName: "Internal Thoughts/Quoted Phrases Punctuation", source: "Entre los impostores", category: "typography" },
  { englishName: "Text Alignment", source: "Entre los impostores", category: "format" },
  { englishName: "Punctuation: Spacing", source: "Entre los impostores", category: "typography" },
  { englishName: "Adjective Choice for 'digno'", source: "Conservatismo", category: "style" },
  { englishName: "Title Page Structure", source: "Nosotros", category: "format" },
  { englishName: "Chapter Titles: Numbering", source: "Entre los impostores", category: "format" },
  { englishName: "Text Justification", source: "Entre los impostores", category: "format" },
  { englishName: "Typo Correction: Missing 'r'", source: "Conservatismo", category: "grammar" },
  { englishName: "Spelling Correction: 'ealeza'", source: "Conservatismo", category: "grammar" },
  { englishName: "Page Numbering", source: "Nosotros", category: "format" },
  { englishName: "Use of Em-dashes for Parenthetical Phrases", source: "Conservatismo", category: "typography" },
  { englishName: "Page Number Placement", source: "Nosotros", category: "format" },
  { englishName: "Dedication Page", source: "Nosotros", category: "format" },
  { englishName: "Spelling Correction: 'erecho'", source: "Conservatismo", category: "grammar" },
  { englishName: "Font Style", source: "Nosotros", category: "format" },
  { englishName: "Dialogue Punctuation", source: "Conservatismo", category: "typography" },
  { englishName: "Colophon Inclusion", source: "Nosotros", category: "format" },
  { englishName: "First Paragraph Indentation", source: "Entre los impostores", category: "format" },
  { englishName: "Spanish Guillemets for Quotes", source: "Entre los impostores", category: "typography" },
  { englishName: "Spelling Correction: 'onarca'", source: "Conservatismo", category: "grammar" },
  { englishName: "Punctuation: Colons", source: "Entre los impostores", category: "typography" },
  { englishName: "Capitalization: 'Humanidad'", source: "Conservatismo", category: "grammar" },
  { englishName: "Chapter Title Formatting", source: "Conservatismo", category: "format" },
  { englishName: "Ellipsis Character", source: "Nosotros", category: "typography" },
  { englishName: "Consistency in Noun Usage", source: "Entre los impostores", category: "style" },
  { englishName: "Capitalization After Dialogue Tag", source: "Entre los impostores", category: "grammar" },
  { englishName: "Footnote Usage", source: "Conservatismo", category: "format" },
  { englishName: "Chapter Title Formatting", source: "Nosotros", category: "format" },
  { englishName: "Grade Level Terminology", source: "Conservatismo", category: "style" },
  { englishName: "Em Dashes for Parenthetical Phrases", source: "Nosotros", category: "typography" },
  { englishName: "Bibliography Format", source: "Entre los impostores", category: "format" },
  { englishName: "Capitalization of Proper Nouns", source: "Conservatismo", category: "grammar" },
  { englishName: "Dedication Page Formatting", source: "Nosotros", category: "format" },
  // Extra from other runs  
  { englishName: "List Item Markers", source: "Nosotros", category: "format" },
  { englishName: "Ellipses Formatting", source: "Conservatismo", category: "typography" },
  { englishName: "Possessive Pronoun Usage", source: "Entre los impostores", category: "grammar" },
];

// Fix typo in source data
const RULES_FIXED = DELETED_RULES.map(r => ({
  englishName: r.englishName ?? r.englipshName,
  source: r.source,
  category: r.category || "style",
})).filter(r => r.englishName);

async function translateRuleBatch(batch) {
  const list = batch.map((r, i) => `${i + 1}. "${r.englishName}" (categoría: ${r.category})`).join("\n");
  const prompt = `Eres un editor experto en lengua española. Para cada norma editorial, 
proporciona en español:
- "nombre": nombre breve y claro de la norma (máx. 8 palabras)
- "descripcion": explicación práctica de la norma (1-2 frases)

Lista de normas a traducir/adaptar al español:
${list}

Devuelve un array JSON con exactamente ${batch.length} objetos con campos "nombre" y "descripcion".
IMPORTANTE: Toda la respuesta debe estar en español.`;

  try {
    const response = await ai.models.generateContent({
      model: "gemini-2.0-flash",
      contents: prompt,
      config: {
        responseMimeType: "application/json",
        temperature: 0.2,
      },
    });
    const text = response.text?.trim() ?? "[]";
    return JSON.parse(text);
  } catch (e) {
    console.error("  ⚠️  Error Gemini:", e.message);
    return batch.map(r => ({
      nombre: r.englishName,
      descripcion: `Norma aplicada en ${r.source}.`,
    }));
  }
}

async function main() {
  console.log(`\n🌐 Restaurando ${RULES_FIXED.length} normas en español para: ${ORG_NAME}\n`);

  const BATCH_SIZE = 10;
  const orgRef = db.collection("organizations").doc(ORG_ID);
  let totalInserted = 0;

  for (let i = 0; i < RULES_FIXED.length; i += BATCH_SIZE) {
    const batch = RULES_FIXED.slice(i, i + BATCH_SIZE);
    console.log(`📦 Procesando ${i + 1}-${Math.min(i + BATCH_SIZE, RULES_FIXED.length)}/${RULES_FIXED.length}...`);

    const translated = await translateRuleBatch(batch);

    if (!DRY_RUN) {
      const firestoreBatch = db.batch();
      for (let j = 0; j < batch.length; j++) {
        const t = translated[j] ?? { nombre: batch[j].englishName, descripcion: "" };
        const ruleId = `restored_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`;
        const ruleRef = orgRef.collection("rules").doc(ruleId);
        firestoreBatch.set(ruleRef, {
          id: ruleId,
          name: t.nombre,
          rule: t.nombre,
          description: t.descripcion,
          category: batch[j].category,
          source: batch[j].source,
          status: "active",
          createdAt: FieldValue.serverTimestamp(),
        });
        console.log(`  ✅ "${batch[j].englishName}" → "${t.nombre}"`);
        totalInserted++;
      }
      await firestoreBatch.commit();
    } else {
      for (let j = 0; j < batch.length; j++) {
        const t = translated[j] ?? { nombre: batch[j].englishName, descripcion: "" };
        console.log(`  [dry-run] "${batch[j].englishName}" → "${t.nombre}"`);
      }
    }
    
    // Rate limit: 1s between batches
    await new Promise(r => setTimeout(r, 1000));
  }

  console.log(`\n✅ Total restauradas: ${totalInserted}`);
  process.exit(0);
}

main().catch(e => { console.error(e); process.exit(1); });

File: ai-worker/scripts/restore_rules_spanish.py
"""
restore_rules_spanish.py
Regenera en español las normas editoriales que se eliminaron.
Usa el cliente Gemini ya configurado en el ai-worker.

Uso: cd ai-worker && python3 scripts/restore_rules_spanish.py [--dry-run]
"""

import sys
import os
import json
import time
import uuid
import argparse

# ── Setup path ─────────────────────────────────────────────
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Cargar settings y Firebase igual que en main.py
import firebase_admin
from firebase_admin import credentials, firestore as fs

sa_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "service-account.json")
if not firebase_admin._apps:
    cred = credentials.Certificate(sa_path)
    firebase_admin.initialize_app(cred)

db = fs.client()

# ── Gemini ──────────────────────────────────────────────────
try:
    from google import genai
    from google.genai import types

    # GCP Application Default Credentials (works since service-account.json gives SA)
    api_key = os.environ.get("GEMINI_API_KEY", "")
    if not api_key:
        # Try to read from ai-worker .env
        env_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), ".env")
        if os.path.exists(env_path):
            for line in open(env_path):
                if line.startswith("GEMINI_API_KEY="):
                    api_key = line.strip().split("=", 1)[1].strip('"\'')
                    break

    if not api_key:
        print("❌ GEMINI_API_KEY no encontrada. Ejecuta: export GEMINI_API_KEY=tu_clave")
        sys.exit(1)

    client = genai.Client(api_key=api_key)
    print(f"✅ Gemini client configurado")
except Exception as e:
    print(f"❌ Error configurando Gemini: {e}")
    sys.exit(1)

# ── Org ──────────────────────────────────────────────────────
ORG_ID = "G8rTqFMuH1xFACIsjEsV"
ORG_NAME = "Biblioteca Homo Legens"

# ── Lista de normas eliminadas (del dry-run) ─────────────────
DELETED_RULES = [
    {"name": "Italics: Foreign words and Latin phrases", "source": "Nosotros", "category": "typography"},
    {"name": "Copyright Page Inclusion", "source": "Nosotros", "category": "format"},
    {"name": "Quotation Marks: Primary", "source": "Nosotros", "category": "typography"},
    {"name": "Thousand Separator for Years", "source": "Nosotros", "category": "style"},
    {"name": "Consistent Numbering Style", "source": "Nosotros", "category": "style"},
    {"name": "Italics for Foreign Words and Titles", "source": "Conservatismo", "category": "typography"},
    {"name": "Capitalization of Religious Terms", "source": "Conservatismo", "category": "grammar"},
    {"name": "Font Style", "source": "Conservatismo", "category": "format"},
    {"name": "Spelling Correction: 'umanidad'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Chapter Headings Alignment", "source": "Entre los impostores", "category": "format"},
    {"name": "Chapter Title Formatting", "source": "Entre los impostores", "category": "format"},
    {"name": "Consistent Spacing", "source": "Entre los impostores", "category": "format"},
    {"name": "Verb Choice for 'marcha'", "source": "Conservatismo", "category": "style"},
    {"name": "Chapter Title Formatting", "source": "Nosotros", "category": "format"},
    {"name": "Final Page Inclusion", "source": "Nosotros", "category": "format"},
    {"name": "Em Dashes for Parenthetical Phrases", "source": "Conservatismo", "category": "typography"},
    {"name": "Block Quotes for Extended Passages", "source": "Entre los impostores", "category": "format"},
    {"name": "Punctuation: Guillemets for quotes", "source": "Nosotros", "category": "typography"},
    {"name": "Spelling Correction: 'aqullos'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Punctuation: Em dashes", "source": "Conservatismo", "category": "typography"},
    {"name": "Consistency in Naming Conventions", "source": "Entre los impostores", "category": "style"},
    {"name": "Em-Dashes for Parentheticals", "source": "Entre los impostores", "category": "typography"},
    {"name": "Noun Choice for 'cole'", "source": "Conservatismo", "category": "style"},
    {"name": "Colophon Formatting", "source": "Entre los impostores", "category": "format"},
    {"name": "Spelling Correction: 'istoria'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Sentence Structure and Flow", "source": "Entre los impostores", "category": "style"},
    {"name": "English Double Quotes for Secondary Quotes", "source": "Nosotros", "category": "typography"},
    {"name": "Consistency in Acronyms/Abbreviations", "source": "Conservatismo", "category": "style"},
    {"name": "Capitalization: 'Estado'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Font Style: Body Text", "source": "Entre los impostores", "category": "format"},
    {"name": "Footnote Style", "source": "Entre los impostores", "category": "format"},
    {"name": "Ellipsis Spacing", "source": "Entre los impostores", "category": "typography"},
    {"name": "Spelling Correction: 'rogresismo'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Typo Correction: Missing 'd' or 'g'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Imprint Page Details", "source": "Nosotros", "category": "format"},
    {"name": "Table of Contents Inclusion", "source": "Nosotros", "category": "format"},
    {"name": "Comma Usage for Clarity", "source": "Entre los impostores", "category": "grammar"},
    {"name": "Use of Spanish Guillemets for Quotes", "source": "Conservatismo", "category": "typography"},
    {"name": "English Single Quotes for Tertiary Quotes", "source": "Nosotros", "category": "typography"},
    {"name": "Logo Placement", "source": "Entre los impostores", "category": "format"},
    {"name": "Page Numbering", "source": "Entre los impostores", "category": "format"},
    {"name": "Comma Usage for Parenthetical Phrases", "source": "Conservatismo", "category": "grammar"},
    {"name": "Footnote Numbering", "source": "Entre los impostores", "category": "format"},
    {"name": "Punctuation: Semicolons", "source": "Nosotros", "category": "typography"},
    {"name": "Punctuation: Periods", "source": "Entre los impostores", "category": "typography"},
    {"name": "Legal Disclaimer", "source": "Nosotros", "category": "format"},
    {"name": "Quotation Marks: Nested/Specific Terms", "source": "Conservatismo", "category": "typography"},
    {"name": "Dialogue Punctuation A", "source": "Entre los impostores", "category": "typography"},
    {"name": "Dialogue Punctuation B", "source": "Nosotros", "category": "typography"},
    {"name": "Em Dash Usage in Dialogue", "source": "Conservatismo", "category": "typography"},
    {"name": "Content Expansion/Clarification", "source": "Entre los impostores", "category": "style"},
    {"name": "Quotation Marks Usage", "source": "Conservatismo", "category": "typography"},
    {"name": "Paragraph Indentation", "source": "Entre los impostores", "category": "format"},
    {"name": "Pronoun Agreement", "source": "Conservatismo", "category": "grammar"},
    {"name": "Punctuation: Ellipses", "source": "Nosotros", "category": "typography"},
    {"name": "Typo Correction: 'onservatismo'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Internal Thoughts Punctuation", "source": "Entre los impostores", "category": "typography"},
    {"name": "Text Alignment", "source": "Entre los impostores", "category": "format"},
    {"name": "Punctuation: Spacing", "source": "Entre los impostores", "category": "typography"},
    {"name": "Adjective Choice for 'digno'", "source": "Conservatismo", "category": "style"},
    {"name": "Title Page Structure", "source": "Nosotros", "category": "format"},
    {"name": "Chapter Titles: Numbering", "source": "Entre los impostores", "category": "format"},
    {"name": "Text Justification", "source": "Entre los impostores", "category": "format"},
    {"name": "Typo Correction: Missing 'r'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Spelling Correction: 'ealeza'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Page Numbering (portada)", "source": "Nosotros", "category": "format"},
    {"name": "Use of Em-dashes for Parenthetical Phrases", "source": "Conservatismo", "category": "typography"},
    {"name": "Page Number Placement", "source": "Nosotros", "category": "format"},
    {"name": "Dedication Page", "source": "Nosotros", "category": "format"},
    {"name": "Spelling Correction: 'erecho'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Font Style (Nosotros)", "source": "Nosotros", "category": "format"},
    {"name": "Dialogue Punctuation C", "source": "Conservatismo", "category": "typography"},
    {"name": "Colophon Inclusion", "source": "Nosotros", "category": "format"},
    {"name": "First Paragraph Indentation", "source": "Entre los impostores", "category": "format"},
    {"name": "Spanish Guillemets for Quotes", "source": "Entre los impostores", "category": "typography"},
    {"name": "Spelling Correction: 'onarca'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Punctuation: Colons", "source": "Entre los impostores", "category": "typography"},
    {"name": "Capitalization: 'Humanidad'", "source": "Conservatismo", "category": "grammar"},
    {"name": "Chapter Title Formatting (Conservatismo)", "source": "Conservatismo", "category": "format"},
    {"name": "Ellipsis Character", "source": "Nosotros", "category": "typography"},
    {"name": "Consistency in Noun Usage", "source": "Entre los impostores", "category": "style"},
    {"name": "Capitalization After Dialogue Tag", "source": "Entre los impostores", "category": "grammar"},
    {"name": "Footnote Usage", "source": "Conservatismo", "category": "format"},
    {"name": "Chapter Title Formatting (Nosotros B)", "source": "Nosotros", "category": "format"},
    {"name": "Grade Level Terminology", "source": "Conservatismo", "category": "style"},
    {"name": "Em Dashes for Parenthetical Phrases (Nosotros)", "source": "Nosotros", "category": "typography"},
    {"name": "Bibliography Format", "source": "Entre los impostores", "category": "format"},
    {"name": "Capitalization of Proper Nouns", "source": "Conservatismo", "category": "grammar"},
    {"name": "Dedication Page Formatting", "source": "Nosotros", "category": "format"},
    {"name": "List Item Markers", "source": "Nosotros", "category": "format"},
    {"name": "Ellipses Formatting", "source": "Conservatismo", "category": "typography"},
    {"name": "Possessive Pronoun Usage", "source": "Entre los impostores", "category": "grammar"},
]


def translate_batch(batch: list) -> list:
    """Translate a batch of English rule names to Spanish using Gemini."""
    list_text = "\n".join(
        f'{i+1}. "{r["name"]}" (fuente: {r["source"]}, categoría: {r["category"]})'
        for i, r in enumerate(batch)
    )
    prompt = f"""Eres un editor experto en lengua española. 
Para cada norma editorial de la siguiente lista, proporciona en español:
- "nombre": nombre breve y claro de la norma editorial (máx. 8 palabras, en español)
- "descripcion": explicación práctica concisa de la norma (1-2 frases, en español)

Lista de normas (originalmente en inglés por un error del sistema, ahora deben traducirse):
{list_text}

Devuelve exactamente un array JSON con {len(batch)} objetos.
Cada objeto debe tener los campos "nombre" y "descripcion" en español.
IMPORTANTE: Toda la respuesta debe estar en español. No uses inglés en ningún campo."""

    schema = {
        "type": "ARRAY",
        "items": {
            "type": "OBJECT",
            "properties": {
                "nombre": {"type": "STRING"},
                "descripcion": {"type": "STRING"},
            },
            "required": ["nombre", "descripcion"],
        },
    }

    try:
        response = client.models.generate_content(
            model="gemini-2.0-flash",
            contents=prompt,
            config=types.GenerateContentConfig(
                response_mime_type="application/json",
                response_schema=schema,
                temperature=0.2,
            ),
        )
        return json.loads(response.text)
    except Exception as e:
        print(f"  ⚠️  Error Gemini: {e}")
        # Fallback: return minimal Spanish names
        return [{"nombre": r["name"], "descripcion": f"Norma aplicada en {r['source']}."} for r in batch]


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--dry-run", action="store_true")
    args = parser.parse_args()

    dry = args.dry_run
    print(f"\n🌐 Restaurando {len(DELETED_RULES)} normas en español para: {ORG_NAME}")
    if dry:
        print("   (DRY RUN — no se escribirá nada en Firestore)\n")

    org_ref = db.collection("organizations").document(ORG_ID)
    BATCH_SIZE = 10
    total = 0

    for i in range(0, len(DELETED_RULES), BATCH_SIZE):
        batch = DELETED_RULES[i:i + BATCH_SIZE]
        end = min(i + BATCH_SIZE, len(DELETED_RULES))
        print(f"📦 Procesando {i+1}-{end}/{len(DELETED_RULES)}...")

        translated = translate_batch(batch)

        if not dry:
            firestore_batch = db.batch()
            for j, rule in enumerate(batch):
                t = translated[j] if j < len(translated) else {"nombre": rule["name"], "descripcion": ""}
                rule_id = f"restored_{uuid.uuid4().hex[:10]}"
                rule_ref = org_ref.collection("rules").document(rule_id)
                firestore_batch.set(rule_ref, {
                    "id": rule_id,
                    "name": t["nombre"],
                    "rule": t["nombre"],
                    "description": t["descripcion"],
                    "category": rule["category"],
                    "source": rule["source"],
                    "status": "active",
                    "createdAt": fs.SERVER_TIMESTAMP,
                })
                print(f"  ✅ \"{rule['name']}\" → \"{t['nombre']}\"")
                total += 1
            firestore_batch.commit()
        else:
            for j, rule in enumerate(batch):
                t = translated[j] if j < len(translated) else {"nombre": rule["name"], "descripcion": ""}
                print(f"  [dry] \"{rule['name']}\" → \"{t['nombre']}\"")

        time.sleep(1)  # rate limit

    print(f"\n✅ Total restauradas: {total}")


if __name__ == "__main__":
    main()
